pkg/controller/storagecluster/cephcluster_test.go

```diff
@@ -262,5 +262,84 @@ func TestStorageClassDeviceSetCreation(t *testing.T) {
 		}
 
 	}
+	sc4 := &api.StorageCluster{}
+	sc4.Spec.StorageDeviceSets = mockDeviceSets
+	sc4.Status.NodeTopologies = &api.NodeTopologyMap{
+		Labels: map[string]api.TopologyLabelValues{
+			defaults.RackTopologyKey: []string{
+				"rack1",
+				"rack2",
+				"rack3",
+			},
+		},
+	}
+	cases = []struct {
+		label                string
+		sc                   *api.StorageCluster
+		topologyKey          string
+		lenOfMatchExpression int
+	}{
+		{
+			label:                "case 1",
+			sc:                   sc1,
+			topologyKey:          "zone",
+			lenOfMatchExpression: 1,
+		},
+		{
+			label:                "case 2",
+			sc:                   sc2,
+			topologyKey:          "zone",
+			lenOfMatchExpression: 1,
+		},
+		{
+			label:                "case 3",
+			sc:                   sc3,
+			topologyKey:          "zone",
+			lenOfMatchExpression: 0,
+		},
+		{
+			label:                "case 3",
+			sc:                   sc4,
+			topologyKey:          "rack",
+			lenOfMatchExpression: 1,
+		},
+	}
+	// Testing StorageClassDeviceSetCreation for kube version 1.19 and above
+	serverVersion = &version.Info{
+		Major: defaults.KubeMajorTSC,
+		Minor: defaults.KubeMinorTSC,
+	}
+	for _, c := range cases {
+
+		actual := newStorageClassDeviceSets(c.sc, serverVersion)
+		assert.Equal(t, defaults.DeviceSetReplica, len(actual))
+		deviceSet := c.sc.Spec.StorageDeviceSets[0]
+
+		for i, scds := range actual {
+			assert.Equal(t, fmt.Sprintf("%s-%d", deviceSet.Name, i), scds.Name)
+			assert.Equal(t, deviceSet.Count/3, scds.Count)
+			assert.Equal(t, defaults.DaemonResources["osd"], scds.Resources)
+			assert.Equal(t, deviceSet.DataPVCTemplate, scds.VolumeClaimTemplates[0])
+			assert.Equal(t, true, scds.Portable)
+			assert.Equal(t, c.sc.Spec.Encryption.Enable, scds.Encrypted)
+			assert.Equal(t, getPlacement(c.sc, "osd"), scds.Placement)
+			if c.lenOfMatchExpression == 0 {
+				assert.Nil(t, scds.Placement.NodeAffinity)
+			} else {
+			matchExpressions := scds.Placement.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions
+			assert.Equal(t, c.lenOfMatchExpression, len(matchExpressions))
+			}
+			topologyKey := scds.PreparePlacement.TopologySpreadConstraints[0].TopologyKey
+
+			if c.topologyKey == "rack" {
+				assert.Equal(t, defaults.RackTopologyKey, topologyKey)
```

Comment:
`c.topologyKey` is used for filtering among the cases defined, when it is of rack the topologyKey is ensured to hold the value `topology.rook.io/rack` instead of `failure-domain.beta.kubernetes.io/zone`

Reply:
...ah! Gotcha. Thanks!